buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchJavaPlugin.java

```diff
@@ -0,0 +1,378 @@
+     * direct dependency version. This specially named and unique configuration
+     * will contain all of the transitive dependencies of this particular
+     * dependency. We can then use this configuration during pom generation
+     * to iterate the transitive dependencies and add excludes.
+     */
+    public static void configureConfigurations(Project project) {
+        // we want to test compileOnly deps!
+        Configuration compileOnlyConfig = project.getConfigurations().getByName(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        Configuration compileConfig = project.getConfigurations().getByName(JavaPlugin.TEST_COMPILE_CONFIGURATION_NAME);
+        compileConfig.extendsFrom(compileOnlyConfig);
+
+        // we are not shipping these jars, we act like dumb consumers of these things
+        if (project.getPath().startsWith(":test:fixtures") || project.getPath().equals(":build-tools")) {
+            return;
+        }
+        // fail on any conflicting dependency versions
+        project.getConfigurations().all(configuration -> {
+            if (configuration.getName().endsWith("Fixture")) {
+                // just a self contained test-fixture configuration, likely transitive and hellacious
+                return;
+            }
+            configuration.resolutionStrategy(ResolutionStrategy::failOnVersionConflict);
+        });
+
+        // force all dependencies added directly to compile/testCompile to be non-transitive, except for ES itself
+        Consumer<String> disableTransitiveDeps = configName -> {
+            Configuration config = project.getConfigurations().getByName(configName);
+            config.getDependencies().all(dep -> {
+                if (dep instanceof ModuleDependency
+                    && dep instanceof ProjectDependency == false
+                    && dep.getGroup().startsWith("org.elasticsearch") == false) {
+                    ((ModuleDependency) dep).setTransitive(false);
+                }
+            });
+        };
+        disableTransitiveDeps.accept(JavaPlugin.COMPILE_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.TEST_COMPILE_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+        disableTransitiveDeps.accept(JavaPlugin.RUNTIME_ONLY_CONFIGURATION_NAME);
+    }
+
+    /** Adds compiler settings to the project */
+    public static void configureCompile(Project project) {
+        ExtraPropertiesExtension ext = project.getExtensions().getByType(ExtraPropertiesExtension.class);
+        ext.set("compactProfile", "full");
+
+        JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
+        java.setSourceCompatibility(BuildParams.getMinimumRuntimeVersion());
+        java.setTargetCompatibility(BuildParams.getMinimumRuntimeVersion());
+
+        Function<File, String> canonicalPath = file -> {
```

Comment:
This seems like something we'd want to throw in one of our util classes.

Reply:
Possibly, but I'd rather add it there once we have more uses.